src/test/java/org/xbill/DNS/LookupTest.java

```diff
@@ -2,89 +2,345 @@
 package org.xbill.DNS;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
 import static org.xbill.DNS.ResolverConfig.CONFIGPROVIDER_SKIP_INIT;
 
 import java.io.IOException;
+import java.io.InterruptedIOException;
 import java.net.InetAddress;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
 import org.mockito.Mockito;
-import org.mockito.stubbing.Answer;
 
 public class LookupTest {
+  public static final Name DUMMY_NAME = Name.fromConstantString("ignored.");
+  private Resolver mockResolver;
+
+  @BeforeEach
+  void before() {
+    System.setProperty(CONFIGPROVIDER_SKIP_INIT, "true");
+    mockResolver = Mockito.mock(Resolver.class);
+  }
+
+  @AfterEach
+  void after() {
+    System.clearProperty(CONFIGPROVIDER_SKIP_INIT);
+  }
+
+  @Test
+  void testRun_absoluteQuery() throws Exception {
+    wireUpMockResolver(mockResolver, this::simpleAnswer);
+
+    Record[] results = makeLookupWithResolver(mockResolver, "example.com.").run();
+
+    ArgumentCaptor<Message> messageCaptor = ArgumentCaptor.forClass(Message.class);
+    verify(mockResolver).send(messageCaptor.capture());
+
+    assertEquals(
+        Record.newRecord(Name.fromConstantString("example.com."), Type.A, DClass.IN, 0L),
+        messageCaptor.getValue().getSection(Section.QUESTION).get(0));
+
+    assertEquals(1, results.length);
+  }
+
+  @Test
+  void testRun_relativeQueryIsMadeAbsolute() throws Exception {
+    wireUpMockResolver(mockResolver, this::simpleAnswer);
+
+    Record[] results = makeLookupWithResolver(mockResolver, "example.com").run();
+
+    ArgumentCaptor<Message> messageCaptor = ArgumentCaptor.forClass(Message.class);
+    verify(mockResolver).send(messageCaptor.capture());
+
+    assertEquals(
+        Record.newRecord(Name.fromConstantString("example.com."), Type.A, DClass.IN, 0L),
+        messageCaptor.getValue().getSection(Section.QUESTION).get(0));
+
+    assertEquals(1, results.length);
+  }
+
+  @Test
+  void testRun_searchAppended() throws Exception {
+    wireUpMockResolver(mockResolver, this::simpleAnswer);
+
+    Lookup lookup = makeLookupWithResolver(mockResolver, "host");
+    lookup.setSearchPath("example.com");
+    Record[] results = lookup.run();
+
+    ArgumentCaptor<Message> messageCaptor = ArgumentCaptor.forClass(Message.class);
+    verify(mockResolver).send(messageCaptor.capture());
+
+    assertEquals(
```

Comment:
The default ndots is 1, a lookup for "host" should thus initially be made relative, i.e. without search path. I'm quite surprised this test passed sometimes, but it correctly failed on the Windows test run.

Reply:
That is not how I read Lookup.java:601-603. A Name with the naked value `host` has `labels() == 1` and as such it doesn't match the conditional looking for labels() values more than 1 in the default case.

My understanding of the default value of `ndots` is that most people want A to be interpreted relative to the search path but they want A.B to be interpreted as a fully qualified Name. However, some people might have a more complicated setup, with hosts in `DC_A` and `DC_B`: thus they want to be able to set `ndots` to a higher value and be able to resolve `HOST.DC_A` as something relative to search path.

With this in mind it seems reasonable to expect a lookup for a naked `host` with `ndots == 1` to first be queried with each of the search path appended to it.

Reply:
Yes, you're right. I always get confused by the wording of resolv.conf's manual:

> ndots:n
sets a threshold for the number of dots which must appear in a name given to res_query(3) (see resolver(3)) before an initial absolute query will be made. The default for n is 1, meaning that if there are any dots in a name, the name will be tried first as an absolute name before any search list elements are appended to it. The value for this option is silently capped to 15.

Anyway, something's wrong somewhere (not necessarily your new tests), as the tests fail when run on Windows.

Reply:
I'll dig out an intel laptop and a  Window VM and see if I can reproduce the problem and get to the bottom of this

Reply:
Got an M1 as a parting gift?
If you push directly to dnsjava I can check out your branches more easily.

Reply:
It turns out that the `WindowsResolverConfigProvider` sets ndots to -1 to indicate that it does not support that configuration, so this failure uncovered a real customer facing issue. Yay!